src/test/pages/Contact.test.tsx

```diff
@@ -0,0 +1,228 @@
+import * as actions from "../../app/state/actions";
+import { Contact } from "../../app/components/pages/Contact";
+
+const renderContactUs = ({
+  role = "STUDENT",
+  successfulRequest = true,
+}: {
+  role?: TestUserRole;
+  successfulRequest?: boolean;
+} = {}) => {
+  renderTestEnvironment({
+    role: role,
+    PageComponent: Contact,
+    initialRouteEntries: ["/"],
+    extraEndpoints: [
+      rest.post(API_PATH + "/contact", (req, res, ctx) => {
+        return successfulRequest
+          ? res(ctx.status(200), ctx.json({}))
+          : res(
+              ctx.status(400),
+              ctx.json({
+                responseCode: 400,
+                responseCodeType: "Bad Request",
+                errorMessage: "Missing form details.",
+                bypassGenericSiteErrorPage: false,
+              }),
+            );
+      }),
+    ],
+  });
+};
+
+function setLocation(searchString: string) {
+  const mockLocation = {
+    search: searchString,
+  };
+  Object.defineProperty(window, "location", {
+    value: mockLocation,
+  });
+}
+
+const formFields = {
+  firstName: () => screen.getByRole("textbox", { name: /first name/i }),
+  lastName: () => screen.getByRole("textbox", { name: /last name/i }),
+  email: () => screen.getByRole("textbox", { name: /email address/i }),
+  subject: () => screen.getByRole("textbox", { name: /message subject/i }),
+  message: () => screen.getByRole("textbox", { name: "Message" }),
+};
+
+const contactFormSubmitSpy = jest.spyOn(actions, "submitMessage");
+
+describe("Contact page", () => {
+  const userRoles: TestUserRole[] = [...USER_ROLES, "ANONYMOUS"];
+
+  it.each(userRoles)("renders with form and expected headings for %s user", (role) => {
+    renderContactUs({ role });
+    checkPageTitle("Contact us");
+    const contactForm = screen.getByRole("form");
+    const upcomingEvents = screen.getByRole("heading", {
+      name: /upcoming events/i,
+    });
+    const problemWithSite = screen.getByRole("heading", {
+      name: /problems with the site\?/i,
+    });
+    const followUs = screen.getByRole("heading", {
+      name: /follow us/i,
+    });
+    [contactForm, upcomingEvents, problemWithSite, followUs].forEach((el) => expect(el).toBeInTheDocument());
+  });
+
+  it("renders expected social links", () => {
+    renderContactUs();
+    const contactLinks = screen.getByTestId("contact-links");
+    Object.values(SOCIAL_LINKS).forEach(({ name, href }) => {
+      const link = within(contactLinks).getByRole("link", { name: new RegExp(name, "i") });
+      expect(link).toBeInTheDocument();
+      expect(link).toHaveAttribute("href", href);
+    });
+  });
+
+  it("shows correct fields in form and auto-populates user details when a user is logged in", async () => {
+    renderContactUs();
+    const { firstName } = formFields;
+    const allFormFields = Object.values(formFields).map((field) => field());
+    allFormFields.forEach((el) => expect(el).toBeInTheDocument());
```

Comment:
If `allFormFields` was empty, this whole test will still pass. The contents of this `forEach` (and the next one) would never be executed. This is one of the problems of trying to automate too much in the tests. What makes for good production code does not necessarily makes for good testing code. In the testing space, it's better to favour readability and simplicity. For instance, just by reading this test it's not clear what these form fields are supposed to be. I also appreciate that it's hard to create new tests for existing code that you didn't write.

Reply:
Is it possible for `allFormFields` here to be empty? If it's running each value in formFields it would surely fail if any of the getByRoles didn't work? I appreciate the point on readability however, and have refactored.

Reply:
I see what you mean, yes, you're right, the `getBy` would throw an error.